src/spec.ts

```diff
@@ -319,9 +319,31 @@ export function alwaysNoOcclusion(spec: ExtendedUnitSpec): boolean {
   return vlEncoding.isAggregate(spec.encoding);
 }
 
+function hasFieldDef(fieldDefs: FieldDef[], fieldDef: FieldDef): boolean {
+  for (let i = 0; i < fieldDefs.length; i++) {
+    if (identical(fieldDefs[i], fieldDef)) {
+      return true;
+    }
+  }
+  return false;
+}
+
 export function fieldDefs(spec: ExtendedUnitSpec | LayerSpec): FieldDef[] {
   // TODO: refactor this once we have composition
-  return isLayerSpec(spec) ? vlEncoding.fieldDefs(spec.layers[0].encoding) : vlEncoding.fieldDefs(spec.encoding);
+  if (isLayerSpec(spec)) {
+    let fieldDefsAllLayers: FieldDef[] = [];
```

Comment:
Instead of O(n) check why don't you make a dictionary of FieldDef using stable stringify to produce key?